GameState_Play.cpp

```diff
@@ -159,30 +163,50 @@ void GameState_Play::playerDie() {
     }
 }
 
+//TODO:: move this to entity ground manager?
 void GameState_Play::createGroundEntities()
 {
   // create as many ground entities as can possibly be visible at once.
   int lastX = 0;
+  int eligibleSegments[GROUND_DEFINITION_COUNT];
+  int eligibleCount = 0;
+  int currentDifficultyLevel = groundManager.calculateDifficultyLevel();  
+
+  for (int i = 0; i < GROUND_DEFINITION_COUNT; i++) {
+    int segmentDifficulty = pgm_read_byte(&groundDefinitions[i].difficulty);
+
+      if (segmentDifficulty <= currentDifficultyLevel) {
+          eligibleSegments[eligibleCount++] = i;
+      }
+  }
+
   for (int i = 0; i < MAX_VISIBLE_GROUND_PIECES; i++)
   {
     Entity_Ground* groundEntity = &groundPieces[i];
     groundEntity->init(lastX, 60);
     groundManager.addEntity(groundEntity);
     lastX += GROUND_DEFINITION_SIZE * GROUND_SIZE;
-    groundEntity->setData(
-            i == 0
-            ? &flatGround
-            : &groundDefinitions[random(0, GROUND_DEFINITION_COUNT)]
-      );
+
+    if (i == 0 || eligibleCount == 0) {
+        groundEntity->setData(&flatGround);
+    } else {
+        int randomIndex = eligibleSegments[random(0, eligibleCount)];
+        groundEntity->setData(&groundDefinitions[randomIndex]);
+    }
```

Comment:
not comparing the differences too closely, but this looks pretty similar to the code up in EntityManager where it's picking new random sections.  you may be able to save a little bit of space if the code is shared idk